// rettiwt-api/dist/models/errors/TwitterError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TwitterErrorDetails = exports.TwitterError = void 0;
/**
 * The error thrown by Twitter API.
 *
 * @public
 */
class TwitterError extends Error {
    details;
    message;
    name;
    status;
    /**
     * @param error - The error response received from Twitter.
     */
    constructor(error) {
        super(error.message);
        this.details = ((error.response?.data).errors
            ? (error.response?.data).errors.map((item) => new TwitterErrorDetails(item))
            : [new TwitterErrorDetails(error.response?.data)]).map((item) => item.toJSON());
        this.message = error.message;
        this.name = 'TWITTER_ERROR';
        this.status = error.status ?? 500;
    }
}
exports.TwitterError = TwitterError;
/**
 * The error details.
 *
 * @public
 */
class TwitterErrorDetails {
    code;
    message;
    name;
    type;
    /**
     * @param details - The details of the error.
     */
    constructor(details) {
        this.code = details.code;
        this.message = details.message;
        this.name = details.name;
        this.type = details.kind;
    }
    /**
     * @returns The JSON representation of `this` object.
     */
    toJSON() {
        return {
            code: this.code,
            message: this.message,
            name: this.message,
            type: this.type,
        };
    }
}
exports.TwitterErrorDetails = TwitterErrorDetails;
