// rettiwt-api/dist/enums/Resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceType = void 0;
/**
 * The different types of resources that can be interacted with.
 *
 * @public
 */
var ResourceType;
(function (ResourceType) {
    // LIST
    ResourceType["LIST_MEMBERS"] = "LIST_MEMBERS";
    ResourceType["LIST_TWEETS"] = "LIST_TWEETS";
    // MEDIA
    ResourceType["MEDIA_UPLOAD_APPEND"] = "MEDIA_UPLOAD_APPEND";
    ResourceType["MEDIA_UPLOAD_FINALIZE"] = "MEDIA_UPLOAD_FINALIZE";
    ResourceType["MEDIA_UPLOAD_INITIALIZE"] = "MEDIA_UPLOAD_INITIALIZE";
    // TWEET
    ResourceType["TWEET_DETAILS"] = "TWEET_DETAILS";
    ResourceType["TWEET_DETAILS_ALT"] = "TWEET_DETAILS_ALT";
    ResourceType["TWEET_DETAILS_BULK"] = "TWEET_DETAILS_BULK";
    ResourceType["TWEET_LIKE"] = "TWEET_LIKE";
    ResourceType["TWEET_LIKERS"] = "TWEET_LIKERS";
    ResourceType["TWEET_POST"] = "TWEET_POST";
    ResourceType["TWEET_REPLIES"] = "TWEET_REPLIES";
    ResourceType["TWEET_RETWEET"] = "TWEET_RETWEET";
    ResourceType["TWEET_RETWEETERS"] = "TWEET_RETWEETERS";
    ResourceType["TWEET_SCHEDULE"] = "TWEET_SCHEDULE";
    ResourceType["TWEET_SEARCH"] = "TWEET_SEARCH";
    ResourceType["TWEET_UNLIKE"] = "TWEET_UNLIKE";
    ResourceType["TWEET_UNPOST"] = "TWEET_UNPOST";
    ResourceType["TWEET_UNRETWEET"] = "TWEET_UNRETWEET";
    ResourceType["TWEET_UNSCHEDULE"] = "TWEET_UNSCHEDULE";
    // USER
    ResourceType["USER_AFFILIATES"] = "USER_AFFILIATES";
    ResourceType["USER_BOOKMARKS"] = "USER_BOOKMARKS";
    ResourceType["USER_DETAILS_BY_USERNAME"] = "USER_DETAILS_BY_USERNAME";
    ResourceType["USER_DETAILS_BY_ID"] = "USER_DETAILS_BY_ID";
    ResourceType["USER_DETAILS_BY_IDS_BULK"] = "USER_DETAILS_BY_IDS_BULK";
    ResourceType["USER_FEED_FOLLOWED"] = "USER_FEED_FOLLOWED";
    ResourceType["USER_FEED_RECOMMENDED"] = "USER_FEED_RECOMMENDED";
    ResourceType["USER_FOLLOW"] = "USER_FOLLOW";
    ResourceType["USER_FOLLOWING"] = "USER_FOLLOWING";
    ResourceType["USER_FOLLOWERS"] = "USER_FOLLOWERS";
    ResourceType["USER_HIGHLIGHTS"] = "USER_HIGHLIGHTS";
    ResourceType["USER_LIKES"] = "USER_LIKES";
    ResourceType["USER_MEDIA"] = "USER_MEDIA";
    ResourceType["USER_NOTIFICATIONS"] = "USER_NOTIFICATIONS";
    ResourceType["USER_SUBSCRIPTIONS"] = "USER_SUBSCRIPTIONS";
    ResourceType["USER_TIMELINE"] = "USER_TIMELINE";
    ResourceType["USER_TIMELINE_AND_REPLIES"] = "USER_TIMELINE_AND_REPLIES";
    ResourceType["USER_UNFOLLOW"] = "USER_UNFOLLOW";
})(ResourceType || (exports.ResourceType = ResourceType = {}));
