// rettiwt-api/dist/enums/raw/Analytics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RawAnalyticsMetric = exports.RawAnalyticsGranularity = void 0;
/**
 * Granularity types for analytics.
 *
 * @public
 */
var RawAnalyticsGranularity;
(function (RawAnalyticsGranularity) {
    RawAnalyticsGranularity["DAILY"] = "Daily";
    RawAnalyticsGranularity["WEEKLY"] = "Weekly";
    RawAnalyticsGranularity["MONTHLY"] = "Monthly";
})(RawAnalyticsGranularity || (exports.RawAnalyticsGranularity = RawAnalyticsGranularity = {}));
/**
 * The different types of metric someone can request.
 *
 * @public
 */
var RawAnalyticsMetric;
(function (RawAnalyticsMetric) {
    RawAnalyticsMetric["ENGAGEMENTS"] = "Engagements";
    RawAnalyticsMetric["IMPRESSIONS"] = "Impressions";
    RawAnalyticsMetric["PROFILE_VISITS"] = "ProfileVisits";
    RawAnalyticsMetric["FOLLOWS"] = "Follows";
    RawAnalyticsMetric["VIDEO_VIEWS"] = "VideoViews";
    RawAnalyticsMetric["REPLIES"] = "Replies";
    RawAnalyticsMetric["LIKES"] = "Likes";
    RawAnalyticsMetric["RETWEETS"] = "Retweets";
    RawAnalyticsMetric["MEDIA_VIEWS"] = "MediaViews";
    RawAnalyticsMetric["BOOKMARK"] = "Bookmark";
    RawAnalyticsMetric["SHARE"] = "Share";
})(RawAnalyticsMetric || (exports.RawAnalyticsMetric = RawAnalyticsMetric = {}));
