// rettiwt-api/dist/models/auth/AuthCookie.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthCookie = void 0;
/**
 * The cookie containing the tokens that are used to authenticate against Twitter.
 *
 * @internal
 */
class AuthCookie {
    /* eslint-disable @typescript-eslint/naming-convention */
    auth_token;
    ct0;
    kdt;
    twid;
    /* eslint-enable @typescript-eslint/naming-convention */
    /**
     * @param cookies - The cookie list obtained from the browser.
     */
    constructor(cookies) {
        // Initializing defaults
        this.auth_token = '';
        this.ct0 = '';
        this.kdt = '';
        this.twid = '';
        // Parsing the cookies
        for (const cookie of cookies) {
            if (cookie.name == 'kdt') {
                this.kdt = cookie.value;
            }
            else if (cookie.name == 'twid') {
                this.twid = cookie.value;
            }
            else if (cookie.name == 'ct0') {
                this.ct0 = cookie.value;
            }
            else if (cookie.name == 'auth_token') {
                this.auth_token = cookie.value;
            }
        }
    }
    /**
     * @returns the string representation of 'this' object.
     */
    toString() {
        /** The string representation of 'this' object. */
        let outStr = '';
        // Iterating through the (key, value) pairs of this cookie
        for (const [key, value] of Object.entries(this)) {
            outStr += `${key}=${value};`;
        }
        return outStr;
    }
}
exports.AuthCookie = AuthCookie;
